internal/cloud/backend_state.go

```diff
@@ -81,19 +108,59 @@ func (r *remoteClient) Put(state []byte) error {
 
 	// Create the new state.
 	_, err = r.client.StateVersions.Create(ctx, r.workspace.ID, options)
+
 	if err != nil {
-		r.stateUploadErr = true
-		return fmt.Errorf("Error uploading state: %v", err)
+		r.stateUpload.hasErrored = true
+		return fmt.Errorf("error uploading state: %v", err)
 	}
 
 	return nil
 }
 
+func getSchemas(ctxOpts *terraform.ContextOpts, services *disco.Disco, state *states.State) (*terraform.Schemas, error) {
```

Comment:
I'd love to get another opinion on the placement of these operations. It seems like some of this should have been initialized as part of the apply command and I'm wondering how we can leverage the config we already loaded because I assume all this is fairly expensive.

Reply:
All of this data has been initialized earlier as part of the apply command. For example we assign a value to schemas at the `(b *Local) opApply()` level: https://github.com/hashicorp/terraform/blob/79f12b83e10ff3078ef2533833e3b2aa7102f2de/internal/backend/local/backend_apply.go#L83 which calls (*Local) localRun()  
It's within this method that we create a state manager, init the remoteClient and write/persist the state (which is the key operation for uploading state). 
You'd think that maybe throughout this process *terraform.Context could have collected schemas data, but that doesn't seem to be the case. We reload the schemas
at different times, at different places. Though an expensive operation, maybe it's a safe thing?
I'll bring this concern to the core team to get another opinion.